src/Database/LSMTree/Internal/MergingRun.hs

```diff
@@ -130,28 +130,36 @@ new ::
   -> RunFsPaths
   -> V.Vector (Ref (Run m h))
   -> m (Ref (MergingRun m h))
-new hfs hbio resolve caching alloc mergeLevel mergePolicy runPaths runs = do
-    merge <- fromMaybe (error "newMerge: merges can not be empty")
-      <$> Merge.new hfs hbio caching alloc mergeLevel resolve runPaths runs
-    let numInputRuns = NumRuns $ V.length runs
-    let numInputEntries = V.foldMap' Run.size runs
-    spentCreditsVar <- SpentCreditsVar <$> newPrimVar 0
-    unsafeNew mergePolicy numInputRuns numInputEntries MergeMaybeCompleted $
-      OngoingMerge runs spentCreditsVar merge
+new hfs hbio resolve caching alloc mergeLevel mergePolicy runPaths inputRuns = do
+    -- If creating the Merge fails, we must release the reference again.
+    bracketOnError (V.mapM dupRef inputRuns) (V.mapM_ releaseRef) $ \runs -> do
```

Comment:
I mentioned it in https://github.com/IntersectMBO/lsm-tree/pull/498#discussion_r1886597205 too, but this is not safe because each of the `dupRef` operations and that would lead to forgotten references. Similary for `releaseRef`, if one `releaseRef` fails, then it won't try to release the other references

Reply:
Let's continue the conversation there.